enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+
+* The kustomize component loader would be refactored so it can be used for this
+  purpose. We should be able to pre-process kustomize yamls to insert template
+  variables as we do today for core assets.
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovn`, `microshift-storage-topolvm`, `microshift-openshift-router`
+  to preserve the current behavior.
+
+* Convert the `microshift-networking` package would be made a virtual package which
+  depends on `microshift-networking-ovn` to preserver current behavior.
+
+* The `microshift-networking-ovn` package would include the necessary
+  `/usr/lib/microshift/components.d/networking-ovn/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/storage-topolvm` yamls, including crds.
+
+* The `microshift-openshift-router` package would include the necessary
+  `/usr/lib/microshift/components.d/openshift-router` yamls, including crds.
+
+The rebase script should be updated to extract those core components
+as kustomize yamls instead of embedded assets.
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the user histories.
+
+#### Upgrading
+
+Upgrading optional components would behave in the same way as upgrading
+components where the assets are embedded into the MicroShift binary.
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
+when the images are created.
+
+#### Deploying Applications
+
+Applications are deployed as usual, we would add support for multiple
+directories in the user component manifests as described in the 2nd goal.
+
+### API Extensions
+
+None.
+
+### Risks and Mitigations
+
+The service manager complexity is increased, the risk of new bugs would
+be addressed by additional unit testing.
+
+The component combinations would need additional testing handled by
+an increased testing matrix.
+
+Considering networking/storage/router to be independent elements, probably
+an additional set of tests which cover "microshift-core" alone would
+be enough.
+
+### Drawbacks
+
+None.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. 
+### Test Plan
+
+We will run additional test matrix with `microshift-core` alone, in those
+tests any OVN/NetworkPolicy tests would need to be disabled.
+
+Additional unit tests should cover the loading of core components as
+kustomize manifests, as well as the depencency handing between them.
+
+Additional unit tests should cover the loading of multiple `.d` style
+directories with manifests.
+
+## Implementation History
+
+* [openshift/microshift](https://github.com/openshift/microshift)
+* [Design guidelines](https://github.com/openshift/microshift/blob/main/docs/design.md)
+
+## Alternatives
+
```

Comment:
I'm thinking out loud and obviously missing some stuff, but I'm wondering how far this "static assets" (by which I mean yaml and systemd [ovs-init] files, and possible other) approach will get us.

We could think about [Go plugin](https://pkg.go.dev/plugin) being alternative not to RPMs itself, but to the what's inside them.
We'd specify an interface that a plugin needs to implement with all the metadata we need to handle it: version (so it matches MicroShift version), which part is handles ("enum" of CNI, CSI, others) so we could easier order them, it would have assets embedded into the binary and know what kind of assets it has, so it could have some "wait for readiness" logic.

I think that could help avoid some potential code creep into microshift-core.
I can imagine situation that we allow installing two CNIs, but from what I remember (it might have changed) multus NEEDS to start after main CNI, otherwise it won't find it and it wasn't monitoring CNI configs but only once at the start of the container (after which it did nothing because kubelet was hitting the multus bin in the cni dir, not the daemonset).

Having them as dynamic libs would be much easier to scan them, order them, find potential conflicts (that's a little bit of logic in microshift-core) than try to interpret yaml (unless circumvent with additional .json metadata file).
Even going further on that cni+multus example: if master cni needs to be up (the config needs to exist in the `/etc/cni/net.d/`) before multus, then we need to wait before deploying multus.

Reply:
Agree. kustomize approach may allow us to decouple the manifests between microshift-core and microshift plugins, there are other parts we shall consider to decouple from microshift-core, such as logic to parse plugin specific config file, the render/apply of plugin manifests and associated logic (plugin dependency, plugin readiness check, plugin cleanup) etc. ultimitately, I think microshift-core just needs to know which plugin to deploy, the rest will be the same for every plugin from microshift-core perspective.